### ISSUE TITLE
Bounds check not eliminated with enumerator

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#78000]
Consider this program:
```C#
using System;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

internal class Program
{
    static void Main()
    {
        C c = new();
        foreach (int i in c) Use(i);
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static void Use(int i) { }
}

public class C
{
    private int[] _entries;

    public Enumerator GetEnumerator() => new Enumerator(_entries);

    public struct Enumerator : IEnumerator<int>
    {
        private readonly int[] _entries;
        private int _index;

        internal Enumerator(int[] entries)
        {
            _entries = entries;
            _index = -1;
        }

        public bool MoveNext()
        {
            _index++;
            if ((uint)_index < (uint)_entries.Length)
            {
                return true;
            }

            //_index = _entries.Length; // Uncomment this
            return false;
        }

        public readonly int Current => _entries[_index];

        object IEnumerator.Current => Current!;
        void IEnumerator.Reset() => _index = -1;
        void IDisposable.Dispose() { }
    }
}
```
`Main` successfully inlines the enumerator's MoveNext and Current, with the comparison in MoveNext just before returning true enabling the bounds check to be eliminated:
```
; Method Program:Main()
G_M000_IG01:                ;; offset=0000H
       57                   push     rdi
       56                   push     rsi
       53                   push     rbx
       4883EC20             sub      rsp, 32

G_M000_IG02:                ;; offset=0007H
       48B9C0E72B4FFB7F0000 mov      rcx, 0x7FFB4F2BE7C0
       E8EA50B35F           call     CORINFO_HELP_NEWSFAST
       488B7008             mov      rsi, gword ptr [rax+08H]
       BFFFFFFFFF           mov      edi, -1
       8B5E08               mov      ebx, dword ptr [rsi+08H]
       EB0C                 jmp      SHORT G_M000_IG04

G_M000_IG03:                ;; offset=0024H
       8BCF                 mov      ecx, edi
       8B4C8E10             mov      ecx, dword ptr [rsi+4*rcx+10H]
       FF15783C1300         call     [Program:Use(int)]

G_M000_IG04:                ;; offset=0030H
       FFC7                 inc      edi
       3BDF                 cmp      ebx, edi
       77EE                 ja       SHORT G_M000_IG03

G_M000_IG05:                ;; offset=0036H
       4883C420             add      rsp, 32
       5B                   pop      rbx
       5E                   pop      rsi
       5F                   pop      rdi
       C3                   ret      
; Total bytes of code: 62
```
Yay.  But now uncomment the `_index = _entries.Length;` just before the `return false;` in the enumerator.  Even though it doesn't impact the correctness of eliminating the bounds check, the bounds check returns:
```
; Method Program:Main()
G_M000_IG01:                ;; offset=0000H
       57                   push     rdi
       56                   push     rsi
       53                   push     rbx
       4883EC20             sub      rsp, 32

G_M000_IG02:                ;; offset=0007H
       48B9C0E72D4FFB7F0000 mov      rcx, 0x7FFB4F2DE7C0
       E8EA50B15F           call     CORINFO_HELP_NEWSFAST
       488B7008             mov      rsi, gword ptr [rax+08H]
       BFFFFFFFFF           mov      edi, -1
       8B5E08               mov      ebx, dword ptr [rsi+08H]
       EB10                 jmp      SHORT G_M000_IG04

G_M000_IG03:                ;; offset=0024H
       3BFB                 cmp      edi, ebx
       731A                 jae      SHORT G_M000_IG06
       8BCF                 mov      ecx, edi
       8B4C8E10             mov      ecx, dword ptr [rsi+4*rcx+10H]
       FF15743C1300         call     [Program:Use(int)]

G_M000_IG04:                ;; offset=0034H
       FFC7                 inc      edi
       3BDF                 cmp      ebx, edi
       77EA                 ja       SHORT G_M000_IG03

G_M000_IG05:                ;; offset=003AH
       4883C420             add      rsp, 32
       5B                   pop      rbx
       5E                   pop      rsi
       5F                   pop      rdi
       C3                   ret      

G_M000_IG06:                ;; offset=0042H
       E899B9C25F           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3     
; Total bytes of code: 72
```

cc: @EgorBo

**Comment by EgorBo** [comment]
cc @AndyAyersMS - we discussed this in Discord a few days ago. It's that issue with:
1) GT_PHI that references unreachable preds
2) AssertionProps that relies on preds of GT_PHI args while it's not legal to rely on them, they're currently for debug purposes from my understanding.

**Comment by EgorBo** [comment]
Optimistically setting 8.0 milestone for this one, would like to know more about how rangecheck handles GT_PHI + assertions

**Comment by jakobbotsch** [comment]
This was fixed by #94563 + #94699
